fms/modules/attention.py

```diff
@@ -267,6 +273,65 @@ def _sdpa_compute_op(
     return attn
 
 
+def _sdpa_with_sinks_op(
+    query: torch.Tensor,
+    key_cache: torch.Tensor,
+    value_cache: torch.Tensor,
+    nheads: int,
+    kvheads: int,
+    p_dropout: float,
+    scale_factor: Optional[float],
+    **attn_kwargs,
+) -> torch.Tensor:
+    queries = query.transpose(2, 1)
+
+    if key_cache.shape[1] != kvheads and key_cache.shape[2] == kvheads:
+        key_cache = key_cache.transpose(2, 1)
+        value_cache = value_cache.transpose(2, 1)
+
+    # Expand kv so black-box attn will work
+    expansion = nheads // kvheads
+    # k/v: b h l d
+    if expansion != 1:
+        keys_e = key_cache.unsqueeze(2).expand(-1, -1, expansion, -1, -1).flatten(1, 2)
+        values_e = (
+            value_cache.unsqueeze(2).expand(-1, -1, expansion, -1, -1).flatten(1, 2)
+        )
+    else:
+        keys_e = key_cache
+        values_e = value_cache
+
+    attn_sinks = attn_kwargs.get("sinks", None)
+    sliding_window = attn_kwargs.get("sliding_window", None)
+
+    # https://github.com/openai/gpt-oss/blob/main/gpt_oss/torch/model.py#L153
+    # from gpt-oss open ai implementation
+    batch, n_heads, n_tokens, d_head = queries.shape
+    S = attn_sinks.reshape(1, -1, 1, 1).expand(batch, -1, n_tokens, -1)
+
+    mask = torch.triu(queries.new_full((n_tokens, n_tokens), -float("inf")), diagonal=1)
+    if sliding_window and sliding_window > 0:
+        mask += torch.tril(
+            mask.new_full((n_tokens, n_tokens), -float("inf")),
+            diagonal=-sliding_window,
+        )
+    QK = torch.einsum("bhqd,bhkd->bhqk", queries, keys_e)
+
+    if scale_factor is None:
+        scale_factor = 1.0 / math.sqrt(d_head)
+
+    QK *= scale_factor  # type: ignore[arg-type]
+    QK += mask[None, None, :, :]
+    QK = torch.cat([QK, S], dim=-1)
+    QK = QK - QK.max(dim=-1, keepdim=True).values
+    W = torch.softmax(QK, dim=-1)
+    W = W[..., :-1]  # drop the attention sinks after done
+    attn = torch.matmul(W, values_e)
+
+    attn = attn.transpose(2, 1).contiguous()
+    return attn
+
+
```

Comment:
I recommend implementing a fast version of this using FlexAttention for GPUs, it shouldn't be too hard: https://github.com/huggingface/transformers/blob/main/src/transformers/integrations/flex_attention.py#L235 here is the version in HF that includes sinks as well